rig/routing_table/minimise.py

```
@@ -0,0 +1,557 @@
+
+Following these rules ensures that the minimised table will be functionally
+equivalent to the original table provided that the original table was invariant
+under reordering OR was provided in increasing order of generality.
+"""
+from collections import defaultdict, namedtuple
+from rig.routing_table import RoutingTableEntry
+from six import itervalues
+
+
+def minimise(routing_table, target_length=1023):
+    """Reduce the size of a routing table by merging together entries where
+    possible.
+
+    .. warning::
+
+        The minimised table may be larger than `target_length`, it is
+        important to check the size of the returned table.
+
+    .. warning::
+
+        It is assumed that the input routing table is not in any particular
+        order and may be reordered into ascending order of generality without
+        affecting routing correctness.  It is also assumed that if this table
+        is unordered it is at least orthogonal (i.e., there are no two entries
+        which would match the same key) -- this assumption is not checked!
+
+        .. note::
+
+            Use :py:meth:`~rig.routing_table.expand_entries` to generate an
+            orthogonal table and receive warnings if the input table were not
+            orthogonal.
+
+    .. warning::
+
+        The input routing table *must* also include entries which could be
+        removed and replaced by default routing.
+
+    Parameters
+    ----------
+    routing_table : [:py:class:`~rig.routing_table.RoutingTableEntry`, ...]
+        Routing entries to be merged.
+
+    Other Parameters
+    ----------------
+    target_length : int
+        Target length of the routing table; the minimisation procedure will
+        halt once either this target is reached or no further minimisation is
+        possible.
+
+    Returns
+    -------
+    [:py:class:`~rig.routing_table.RoutingTableEntry`, ...]
+        Reduced routing table entries.
```

As with the wrapper, *maybe* this should raise an exception if the target isn't met? Not many applications will have any sensible course of action to take when minimisation fails and forcing everyone to manually check the output length seems like unnecessary boilerplate.